### PR TITLE
Added test to check if charset is set on application/json content type

### PR DESCRIPTION
**Pull request by raam86**
I looks like part of the problem is with that application/json is WithFIxedCharset 
Firstly `application/json` can have utf-16 and utf-32 charsets and secondly I cannot force charsets.

### CODE REVIEW

**Comment by raam86** [comment]
Relevant bug 
https://github.com/akka/akka/issues/19260#issuecomment-166866370

**Comment by ktoso** [comment]
Thanks for the PR, initially I was thinking our rendering could have been wrong, but in light of @jrudolph's findings:

> Here's the bug report for chromium: https://code.google.com/p/chromium/issues/detail?id=438464
and for firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1166558

It indeed seems it is browser bugs... We'll need to think about what to do here. Technically it is not an Akka but, and Akka is abiding the RFC properly it seems.

Pinging @sirthias for feedback if he has time.

**Comment by raam86** [comment]
I am not sure it is wrong per se, however there is an underlying issue where application/json can also have utf-32 charset but it is not possible to set since it's WithFixedCharset.

**Comment by sirthias** [comment]
The current JSON spec says [this](http://tools.ietf.org/html/rfc7158#section-8.1):

> JSON text SHALL be encoded in UTF-8, UTF-16, or UTF-32.  The default
> encoding is UTF-8, and JSON texts that are encoded in UTF-8 are
> interoperable in the sense that they will be read successfully by the
> maximum number of implementations; there are many implementations
> that cannot successfully read texts in other encodings (such as
> UTF-16 and UTF-32).

So, while other unicode encoding are allowed we've not seen any actual use of these in all of spray's (akka-http's) life so far. That's why we went for the fixed UTF-8 encoding because that is the one that people will want to use in 99,999% of the time.

Given the browser pecularities we've decided to render an explicit `charset` parameter within the Content-Type, so that there are no ambiguities (even in the presence of incorrect default charsets on the peer's side). So, I'd say that akka-http does the right thing with regard to JSON de-/encoding at this point.

As to this PR, I think @raam86 is right in improving test coverage here. We should indeed add a test demonstrating that `application/json` content types are rendered in the way we want.
However, I could imagine that the `HttpHeaderSpec` would be the better (and easier) place to add it.
Basically we only need one more test case here: https://github.com/akka/akka/blob/release-2.3-dev/akka-http-core/src/test/scala/akka/http/impl/model/parser/HttpHeaderSpec.scala#L194

**Comment by raam86** [comment]
Can one of the repo owners verify this patch?

**Comment by ktoso** [comment]
OK TO TEST

**Comment by jrudolph** [comment]
LGTM

**Comment by ktoso** [comment]
Thanks a lot @raam86!

**Comment by ktoso** [comment]
LGTM

**Comment by ktoso** [comment]
I squashed it into one commit and merged manually, thanks again @raam86!
